core/src/main/java/tc/oc/pgm/structure/StructureMatchModule.java

```diff
@@ -0,0 +1,54 @@
+package tc.oc.pgm.structure;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+import tc.oc.pgm.api.match.Match;
+import tc.oc.pgm.api.match.MatchModule;
+import tc.oc.pgm.api.module.exception.ModuleLoadException;
+import tc.oc.pgm.api.player.MatchPlayer;
+import tc.oc.pgm.filters.dynamic.FilterMatchModule;
+
+public class StructureMatchModule implements MatchModule {
+
+  private final DynamicScheduler scheduler;
+  private final Match match;
+  private final List<DynamicDefinition> dynamicDefinitions;
+
+  public StructureMatchModule(Match match, List<DynamicDefinition> dynamicDefinitions) {
+    this.match = match;
+    this.dynamicDefinitions = dynamicDefinitions;
+    this.scheduler =
+        new DynamicScheduler(
+            match,
+            Comparator.comparing(
+                dynamic ->
+                    this.dynamicDefinitions.indexOf(
+                        dynamic.getDefinition()))); // TODO does this sort backwards?
+  }
+
+  @Override
+  public void load() throws ModuleLoadException {
+    List<Dynamic> dynamics =
+        dynamicDefinitions.stream().map(d -> new Dynamic(d, match)).collect(Collectors.toList());
+
+    FilterMatchModule fmm = match.needModule(FilterMatchModule.class);
+    for (Dynamic dynamic : dynamics) {
+      match.getFeatureContext().add(dynamic);
+      final DynamicDefinition dynamicDef = dynamic.getDefinition();
+      fmm.onChange(
+          MatchPlayer.class, // TODO make this scope adjustable
```

Comment:
?

Reply:
this was while we talked about the scope attribute, I guess it makes more sense to add action stuff for structures at this point?

Reply:
I think it may simply be absurd to listen to match player for a structure change, this should be listening on a match scope. One player passes the filter and it places? then another stops matching and it removes? what is this?

It should be using a match scope, if the filter you provide is not able to respond to a match then it simply cannot be used there, this is somewhat related to #1050 , where objective modes were listening for onChange without a scope at all (meaning, it'd use `Filterable.class` as a scope) which broke things, as you could use a filter that wouldn't react well to all filterables (eg: it required a player to check something, and couldn't work with just a match query)